include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -132,1033 +84,1120 @@ class alignment_file_input_format_REMOVEME<format_bam> : alignment_file_input_fo
+    static constexpr std::array<uint8_t, 256> char_to_sam_rank
+    {
+        [] () constexpr
         {
-            // magic BAM string
-            if (!std::ranges::equal(stream_view | views::take_exactly_or_throw(4), std::string_view{"BAM\1"}))
-                throw format_error{"File is not in BAM format."};
+            std::array<uint8_t, 256> ret{};
 
-            int32_t tmp32{};
-            read_field(stream_view, tmp32);
+            using index_t = std::make_unsigned_t<char>;
 
-            if (tmp32 > 0) // header text is present
-                read_header(stream_view | views::take_exactly_or_throw(tmp32)
-                                        | views::take_until_and_consume(is_char<'\0'>),
-                            header,
-                            ref_seqs);
+            // ret['M'] = 0; set anyway by initialization
+            ret[static_cast<index_t>('I')] = 1;
+            ret[static_cast<index_t>('D')] = 2;
+            ret[static_cast<index_t>('N')] = 3;
+            ret[static_cast<index_t>('S')] = 4;
+            ret[static_cast<index_t>('H')] = 5;
+            ret[static_cast<index_t>('P')] = 6;
+            ret[static_cast<index_t>('=')] = 7;
+            ret[static_cast<index_t>('X')] = 8;
 
-            int32_t n_ref;
-            read_field(stream_view, n_ref);
+            return ret;
+        }()
+    };
 
-            for (int32_t ref_idx = 0; ref_idx < n_ref; ++ref_idx)
-            {
-                read_field(stream_view, tmp32); // l_name (length of reference name including \0 character)
+    //!\brief Computes the bin number for a given region [beg, end), copied from the official SAM specifications.
+    static uint16_t reg2bin(int32_t beg, int32_t end) noexcept
+    {
+        --end;
+        if (beg >> 14 == end >> 14) return ((1 << 15) - 1) / 7 + (beg >> 14);
+        if (beg >> 17 == end >> 17) return ((1 << 12) - 1) / 7 + (beg >> 17);
+        if (beg >> 20 == end >> 20) return ((1 << 9)  - 1) / 7 + (beg >> 20);
+        if (beg >> 23 == end >> 23) return ((1 << 6)  - 1) / 7 + (beg >> 23);
+        if (beg >> 26 == end >> 26) return ((1 << 3)  - 1) / 7 + (beg >> 26);
+        return 0;
+    }
 
-                string_buffer.resize(tmp32 - 1);
-                std::ranges::copy_n(std::ranges::begin(stream_view), tmp32 - 1, string_buffer.data()); // copy without \0 character
-                std::ranges::next(std::ranges::begin(stream_view)); // skip \0 character
+    using format_sam_base::read_field; // inherit read_field functions from format_base explicitly
 
-                read_field(stream_view, tmp32); // l_ref (length of reference sequence)
+    /*!\brief Reads a arithmetic field from binary stream by directly reinterpreting the bits.
+     * \tparam stream_view_type  The type of the stream as a view.
+     * \param[in, out] stream_view  The stream view to read from.
+     * \param[in, out] target       An arithmetic value to store the parsed value in.
```

Comment:
Really also in?

Reply:
yep. 
1. I need the exact type of the arithmetic value to trigger to correct std::from_chars overload (it does e.g. range/overflow checks). 
2. I get the arithmetic value by reference in the outer `read()` function anyway so I just pass this on instead of copying the result.

Reply:
ah you are right. it's just an out parameter :see

Reply:
not fixed